tsdb/engine/tsm1/reader.go

```diff
@@ -600,12 +627,76 @@ func (d *indirectIndex) Delete(keys []string) {
 	d.offsets = offsets
 }
 
+func (d *indirectIndex) DeleteRange(keys []string, minTime, maxTime int64) {
+	// No keys, nothing to do
+	if len(keys) == 0 {
+		return
+	}
+
+	// If we're deleting the max time range, just use tombstoning to remove the
+	// key from the offsets slice
+	if minTime == math.MinInt64 && maxTime == math.MaxInt64 {
+		d.Delete(keys)
+		return
+	}
+
+	// Is the range passed in outside of the time range for the file?
+	min, max := d.TimeRange()
+	if minTime > max || maxTime < min {
+		return
+	}
+
+	tombstones := map[string][]TimeRange{}
+	for _, k := range keys {
+
+		// Is the range passed in outside the time range for this key?
+		entries := d.Entries(k)
+		min, max := entries[0].MinTime, entries[len(entries)-1].MaxTime
+		if minTime > max || maxTime < min {
+			continue
+		}
+
+		tombstones[k] = append(tombstones[k], TimeRange{minTime, maxTime})
+	}
+
+	if len(tombstones) == 0 {
+		return
+	}
+
+	d.mu.Lock()
+	for k, v := range tombstones {
+		d.tombstones[k] = append(d.tombstones[k], v...)
+	}
+	d.mu.Unlock()
+
+}
+
+func (d *indirectIndex) TombstoneRange(key string) []TimeRange {
+	d.mu.RLock()
+	r := d.tombstones[key]
+	d.mu.RUnlock()
+	return r
+}
+
 func (d *indirectIndex) Contains(key string) bool {
 	return len(d.Entries(key)) > 0
 }
 
 func (d *indirectIndex) ContainsValue(key string, timestamp int64) bool {
-	return d.Entry(key, timestamp) != nil
+	entry := d.Entry(key, timestamp)
+	if entry != nil {
```

Comment:
[Coding Style] The convention in Go is to return early if possibly, then continue the method without being indented in a block. Changing this condition to `if entry == nil { return false }` would remove the large block here